File: generators/org/jetbrains/jet/generators/injectors/GenerateInjectors.java
/*
 * Copyright 2010-2013 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.jet.generators.injectors;

import com.intellij.openapi.project.Project;
import com.intellij.psi.search.GlobalSearchScope;
import org.jetbrains.jet.codegen.ClassBuilderFactory;
import org.jetbrains.jet.codegen.ClassBuilderMode;
import org.jetbrains.jet.codegen.ClassFileFactory;
import org.jetbrains.jet.codegen.ScriptCodegen;
import org.jetbrains.jet.codegen.intrinsics.IntrinsicMethods;
import org.jetbrains.jet.codegen.state.GenerationState;
import org.jetbrains.jet.codegen.state.JetTypeMapper;
import org.jetbrains.jet.di.DependencyInjectorGenerator;
import org.jetbrains.jet.di.GivenExpression;
import org.jetbrains.jet.di.InjectorForTopDownAnalyzer;
import org.jetbrains.jet.lang.ModuleConfiguration;
import org.jetbrains.jet.lang.descriptors.ModuleDescriptor;
import org.jetbrains.jet.lang.descriptors.SubModuleDescriptor;
import org.jetbrains.jet.lang.psi.JetImportsFactory;
import org.jetbrains.jet.lang.resolve.*;
import org.jetbrains.jet.lang.resolve.calls.CallResolver;
import org.jetbrains.jet.lang.resolve.java.*;
import org.jetbrains.jet.lang.resolve.java.resolver.JavaClassResolver;
import org.jetbrains.jet.lang.resolve.lazy.LazyCodeAnalyzer;
import org.jetbrains.jet.lang.resolve.lazy.ScopeProvider;
import org.jetbrains.jet.lang.resolve.lazy.storage.StorageManager;
import org.jetbrains.jet.lang.types.DependencyClassByQualifiedNameResolverDummyImpl;
import org.jetbrains.jet.lang.types.expressions.ExpressionTypingServices;
import org.jetbrains.jet.lang.types.lang.KotlinBuiltIns;

import java.io.IOException;

// NOTE: After making changes, you need to re-generate the injectors.
//       To do that, you can run either this class, or /build.xml/generateInjectors task
public class GenerateInjectors {

    private GenerateInjectors() {
    }

    public static void main(String[] args) throws IOException {
        generateInjectorForTopDownAnalyzerBasic();
        generateInjectorForTopDownAnalyzerForJvm();
        generateInjectorForJavaDescriptorResolver();
        generateInjectorForTopDownAnalyzerForJs();
        generateMacroInjector();
        generateTestInjector();
        generateInjectorForJavaSemanticServices();
        generateInjectorForJvmCodegen();
        generateInjectorForLazyResolve();
        generateInjectorForBodyResolve();
    }

    private static void generateInjectorForLazyResolve() throws IOException {
        DependencyInjectorGenerator generator = new DependencyInjectorGenerator();
        generator.addParameter(Project.class);
        generator.addParameter(ModuleSourcesManager.class);
        generator.addParameter(LazyCodeAnalyzer.class);
        generator.addPublicField(DescriptorResolver.class);
        generator.addPublicField(ExpressionTypingServices.class);
        generator.addPublicField(TypeResolver.class);
        generator.addPublicField(ScopeProvider.class);
        generator.addPublicField(AnnotationResolver.class);
        generator.addPublicField(QualifiedExpressionResolver.class);
        generator.addPublicField(JetImportsFactory.class);
        generator.generate("compiler/frontend/src", "org.jetbrains.jet.di", "InjectorForLazyResolve", GenerateInjectors.class);
    }

    private static void generateInjectorForTopDownAnalyzerBasic() throws IOException {
        DependencyInjectorGenerator generator = new DependencyInjectorGenerator();
        generateInjectorForTopDownAnalyzerCommon(generator);
        generator.addField(DependencyClassByQualifiedNameResolverDummyImpl.class);
        generator.generate("compiler/frontend/src", "org.jetbrains.jet.di", "InjectorForTopDownAnalyzerBasic", GenerateInjectors.class);
    }

    private static void generateInjectorForTopDownAnalyzerForJs() throws IOException {
        DependencyInjectorGenerator generator = new DependencyInjectorGenerator();
        generateInjectorForTopDownAnalyzerCommon(generator);
        generator.addPublicParameter(ModuleConfiguration.class);
        generator.addField(DependencyClassByQualifiedNameResolverDummyImpl.class);
        generator.generate("js/js.translator/src", "org.jetbrains.jet.di", "InjectorForTopDownAnalyzerForJs", GenerateInjectors.class);
    }

    private static void generateInjectorForTopDownAnalyzerForJvm() throws IOException {
        DependencyInjectorGenerator generator = new DependencyInjectorGenerator();
        generator.implementInterface(InjectorForTopDownAnalyzer.class);
        generateInjectorForTopDownAnalyzerCommon(generator);
        generator.generate("compiler/frontend.java/src", "org.jetbrains.jet.di", "InjectorForTopDownAnalyzerForJvm",
                           GenerateInjectors.class);
    }

    private static void generateInjectorForJavaDescriptorResolver() throws IOException {
        DependencyInjectorGenerator generator = new DependencyInjectorGenerator();

        // Parameters
        generator.addPublicParameter(Project.class);
        generator.addPublicParameter(BindingTrace.class);
        generator.addParameter(JavaClassResolutionFacade.class);
        generator.addParameter(StorageManager.class);
        generator.addParameter(SubModuleDescriptor.class);
        generator.addParameter(GlobalSearchScope.class);

        // Fields
        generator.addField(JavaBridgeConfiguration.class);
        generator.addField(JavaDependencyByQualifiedNameResolver.class);
        generator.addPublicField(JavaDescriptorResolver.class);
        generator.addPublicField(JavaClassResolver.class);
        generator.addPublicField(PsiClassFinderImpl.class);
        generator.addPublicField(JavaPackageFragmentProvider.class);

        generator.generate("compiler/frontend.java/src", "org.jetbrains.jet.di", "InjectorForJavaDescriptorResolver",
                           GenerateInjectors.class);
    }

    private static void generateInjectorForTopDownAnalyzerCommon(DependencyInjectorGenerator generator) {
        // Fields
        generator.addPublicField(TopDownAnalyzer.class);
        generator.addPublicField(TopDownAnalysisContext.class);
        generator.addPublicField(BodyResolver.class);
        generator.addPublicField(ControlFlowAnalyzer.class);
        generator.addPublicField(DeclarationsChecker.class);
        generator.addPublicField(DescriptorResolver.class);
        generator.addField(DependencyClassByQualifiedNameResolverDummyImpl.class); // TODO: scripts use this, btu it doesn't seem right

        // Parameters
        generator.addPublicParameter(Project.class);
        generator.addPublicParameter(TopDownAnalysisParameters.class);
        generator.addPublicParameter(BindingTrace.class);
        generator.addParameter(ModuleSourcesManager.class);
    }

    private static void generateMacroInjector() throws IOException {
        DependencyInjectorGenerator generator = new DependencyInjectorGenerator();

        // Fields
        generator.addPublicField(ExpressionTypingServices.class);

        // Parameters
        generator.addPublicParameter(Project.class);
        generator.addParameter(ModuleSourcesManager.class);

        generator.generate("compiler/frontend/src", "org.jetbrains.jet.di", "InjectorForMacros", GenerateInjectors.class);
    }

    private static void generateTestInjector() throws IOException {
        DependencyInjectorGenerator generator = new DependencyInjectorGenerator();

        // Fields
        generator.addPublicField(DescriptorResolver.class);
        generator.addPublicField(ExpressionTypingServices.class);
        generator.addPublicField(TypeResolver.class);
        generator.addPublicField(CallResolver.class);
        generator.addField(true, KotlinBuiltIns.class, null, new GivenExpression("KotlinBuiltIns.getInstance()"));

        // Parameters
        generator.addPublicParameter(Project.class);
        generator.addParameter(ModuleSourcesManager.class);

        generator.generate("compiler/tests", "org.jetbrains.jet.di", "InjectorForTests", GenerateInjectors.class);
    }

    private static void generateInjectorForJavaSemanticServices() throws IOException {
        DependencyInjectorGenerator generator = new DependencyInjectorGenerator();

        // Fields
        generator.addPublicField(JavaDescriptorResolver.class);
        generator.addField(true, BindingTrace.class, null,
                           new GivenExpression("new org.jetbrains.jet.lang.resolve.BindingTraceContext()"));
        generator.addField(JavaBridgeConfiguration.class);
        generator.addField(JavaDependencyByQualifiedNameResolver.class);
        generator.addPublicField(PsiClassFinderImpl.class);
        generator.addField(false, ModuleDescriptor.class, null,
                           new GivenExpression("new org.jetbrains.jet.lang.descriptors.impl.MutableModuleDescriptor(" +
                                               "org.jetbrains.jet.lang.resolve.name.Name.special(\"<dummy>\"), org.jetbrains.jet.lang.resolve.java.JavaToKotlinClassMap.getInstance())"));

        // Parameters
        generator.addPublicParameter(Project.class);
        generator.addParameter(JavaClassResolutionFacade.class);
        generator.addParameter(StorageManager.class);
        generator.addParameter(SubModuleDescriptor.class);
        generator.addParameter(GlobalSearchScope.class);

        generator.generate("compiler/frontend.java/src", "org.jetbrains.jet.di", "InjectorForJavaSemanticServices",
                           GenerateInjectors.class);
    }

    private static void generateInjectorForJvmCodegen() throws IOException {
        DependencyInjectorGenerator generator = new DependencyInjectorGenerator();

        // Parameters
        generator.addPublicParameter(JetTypeMapper.class);
        generator.addPublicParameter(GenerationState.class);
        generator.addParameter(ClassBuilderFactory.class);
        generator.addPublicParameter(Project.class);

        // Fields
        generator.addField(false, BindingTrace.class, "bindingTrace",
                           new GivenExpression("jetTypeMapper.getBindingTrace()"));
        generator.addField(false, BindingContext.class, "bindingContext",
                           new GivenExpression("bindingTrace.getBindingContext()"));
        generator.addField(false, ClassBuilderMode.class, "classBuilderMode",
                           new GivenExpression("classBuilderFactory.getClassBuilderMode()"));
        generator.addPublicField(ScriptCodegen.class);
        generator.addField(true, IntrinsicMethods.class, "intrinsics", null);
        generator.addPublicField(ClassFileFactory.class);

        generator.generate("compiler/backend/src", "org.jetbrains.jet.di", "InjectorForJvmCodegen", GenerateInjectors.class);
    }

    private static void generateInjectorForBodyResolve() throws IOException {
        DependencyInjectorGenerator generator = new DependencyInjectorGenerator();
        // Fields
        generator.addPublicField(BodyResolver.class);

        // Parameters
        generator.addPublicParameter(Project.class);
        generator.addPublicParameter(TopDownAnalysisParameters.class);
        generator.addPublicParameter(BindingTrace.class);
        generator.addPublicParameter(BodiesResolveContext.class);
        generator.addParameter(ModuleSourcesManager.class);
        generator.generate("compiler/frontend/src", "org.jetbrains.jet.di", "InjectorForBodyResolve", GenerateInjectors.class);
    }
}

File: generators/src/org/jetbrains/jet/generators/injectors/GenerateInjectors.kt
/*
 * Copyright 2010-2014 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.jet.generators.injectors

import com.intellij.openapi.project.Project
import org.jetbrains.jet.context.GlobalContext
import org.jetbrains.jet.context.GlobalContextImpl
import org.jetbrains.jet.lang.descriptors.ModuleDescriptor
import org.jetbrains.jet.lang.descriptors.impl.ModuleDescriptorImpl
import org.jetbrains.jet.lang.resolve.*
import org.jetbrains.jet.lang.resolve.java.*
import org.jetbrains.jet.lang.resolve.java.resolver.*
import org.jetbrains.jet.lang.resolve.java.sam.SamConversionResolverImpl
import org.jetbrains.jet.lang.resolve.kotlin.VirtualFileFinder
import org.jetbrains.jet.lang.resolve.lazy.ResolveSession
import org.jetbrains.jet.lang.resolve.lazy.declarations.DeclarationProviderFactory
import org.jetbrains.jet.lang.types.expressions.ExpressionTypingServices
import org.jetbrains.jet.di.*
import org.jetbrains.jet.lang.types.expressions.ExpressionTypingComponents
import org.jetbrains.jet.lang.types.expressions.ExpressionTypingUtils
import org.jetbrains.jet.lang.resolve.calls.CallResolver
import org.jetbrains.jet.lang.resolve.java.structure.impl.JavaPropertyInitializerEvaluatorImpl
import com.intellij.psi.search.GlobalSearchScope
import org.jetbrains.jet.lang.resolve.java.lazy.ModuleClassResolver
import org.jetbrains.jet.lang.resolve.kotlin.DeserializationComponentsForJava
import org.jetbrains.jet.lang.resolve.java.lazy.SingleModuleClassResolver
import org.jetbrains.jet.lang.resolve.kotlin.VirtualFileFinderFactory
import org.jetbrains.jet.lang.resolve.kotlin.JavaDeclarationCheckerProvider
import org.jetbrains.jet.lang.resolve.lazy.KotlinCodeAnalyzer
import org.jetbrains.jet.context.LazyResolveToken

// NOTE: After making changes, you need to re-generate the injectors.
//       To do that, you can run main in this file.
public fun main(args: Array<String>) {
    for (generator in createInjectorGenerators()) {
        try {
            generator.generate()
        }
        catch (e: Throwable) {
            System.err.println(generator.getOutputFile())
            throw e
        }
    }
}

public fun createInjectorGenerators(): List<DependencyInjectorGenerator> =
        listOf(
                generatorForTopDownAnalyzerBasic(),
                generatorForTopDownAnalyzerForJvm(),
                generatorForJavaDescriptorResolver(),
                generatorForLazyResolveWithJava(),
                generatorForTopDownAnalyzerForJs(),
                generatorForMacro(),
                generatorForTests(),
                generatorForLazyResolve(),
                generatorForBodyResolve(),
                generatorForLazyBodyResolve()
        )

private fun DependencyInjectorGenerator.commonForTopDownAnalyzer() {
    parameter(javaClass<Project>())
    parameter(javaClass<GlobalContext>(), useAsContext = true)
    parameter(javaClass<BindingTrace>())
    publicParameter(javaClass<ModuleDescriptor>(), useAsContext = true)

    publicFields(
            javaClass<TopDownAnalyzer>(),
            javaClass<LazyTopDownAnalyzer>()
    )

    field(javaClass<MutablePackageFragmentProvider>())
}

private fun generatorForTopDownAnalyzerBasic() =
        generator("compiler/frontend/src", "org.jetbrains.jet.di", "InjectorForTopDownAnalyzerBasic") {
            commonForTopDownAnalyzer()
            parameter(javaClass<AdditionalCheckerProvider>())
        }

private fun generatorForLazyBodyResolve() =
        generator("compiler/frontend/src", "org.jetbrains.jet.di", "InjectorForLazyBodyResolve") {
            parameter(javaClass<Project>())
            parameter(javaClass<GlobalContext>(), useAsContext = true)
            parameter(javaClass<KotlinCodeAnalyzer>(), name = "analyzer")
            parameter(javaClass<BindingTrace>())
            parameter(javaClass<AdditionalCheckerProvider>())

            field(javaClass<ModuleDescriptor>(), init = GivenExpression("analyzer.getModuleDescriptor()"), useAsContext = true)

            publicFields(
                    javaClass<LazyTopDownAnalyzer>()
            )
        }

private fun generatorForTopDownAnalyzerForJs() =
        generator("js/js.frontend/src", "org.jetbrains.jet.di", "InjectorForTopDownAnalyzerForJs") {
            commonForTopDownAnalyzer()

            field(javaClass<AdditionalCheckerProvider>(),
                  init = GivenExpression(javaClass<AdditionalCheckerProvider.Empty>().getCanonicalName() + ".INSTANCE$"))
        }

private fun generatorForTopDownAnalyzerForJvm() =
        generator("compiler/frontend.java/src", "org.jetbrains.jet.di", "InjectorForTopDownAnalyzerForJvm") {
            commonForTopDownAnalyzer()

            publicField(javaClass<JavaDescriptorResolver>())
            publicField(javaClass<DeserializationComponentsForJava>())

            field(javaClass<AdditionalCheckerProvider>(),
                  init = GivenExpression(javaClass<JavaDeclarationCheckerProvider>().getName() + ".INSTANCE$"))

            field(javaClass <GlobalSearchScope>(),
                  init = GivenExpression(javaClass<GlobalSearchScope>().getName() + ".allScope(project)"))
            fields(
                    javaClass<JavaClassFinderImpl>(),
                    javaClass<TraceBasedExternalSignatureResolver>(),
                    javaClass<TraceBasedJavaResolverCache>(),
                    javaClass<TraceBasedErrorReporter>(),
                    javaClass<PsiBasedMethodSignatureChecker>(),
                    javaClass<PsiBasedExternalAnnotationResolver>(),
                    javaClass<MutablePackageFragmentProvider>(),
                    javaClass<JavaPropertyInitializerEvaluatorImpl>(),
                    javaClass<SamConversionResolverImpl>(),
                    javaClass<JavaSourceElementFactoryImpl>(),
                    javaClass<SingleModuleClassResolver>(),
                    javaClass<JavaFlexibleTypeCapabilitiesProvider>()
            )
            field(javaClass<VirtualFileFinder>(), init = GivenExpression(javaClass<VirtualFileFinder>().getName() + ".SERVICE.getInstance(project)"))
        }

private fun generatorForJavaDescriptorResolver() =
        generator("compiler/frontend.java/src", "org.jetbrains.jet.di", "InjectorForJavaDescriptorResolver") {
            parameters(
                    javaClass<Project>(),
                    javaClass<BindingTrace>()
            )

            publicField(javaClass<GlobalContextImpl>(), useAsContext = true,
                        init = GivenExpression("org.jetbrains.jet.context.ContextPackage.GlobalContext()"))
            publicField(javaClass<ModuleDescriptorImpl>(), name = "module",
                        init = GivenExpression(javaClass<TopDownAnalyzerFacadeForJVM>().getName() + ".createJavaModule(\"<fake-jdr-module>\")"))
            publicField(javaClass<JavaDescriptorResolver>())
            publicField(javaClass<JavaClassFinderImpl>())

            field(javaClass <GlobalSearchScope>(),
                  init = GivenExpression(javaClass<GlobalSearchScope>().getName() + ".allScope(project)"))

            fields(
                    javaClass<TraceBasedExternalSignatureResolver>(),
                    javaClass<TraceBasedJavaResolverCache>(),
                    javaClass<TraceBasedErrorReporter>(),
                    javaClass<PsiBasedMethodSignatureChecker>(),
                    javaClass<PsiBasedExternalAnnotationResolver>(),
                    javaClass<JavaPropertyInitializerEvaluatorImpl>(),
                    javaClass<SamConversionResolverImpl>(),
                    javaClass<JavaSourceElementFactoryImpl>(),
                    javaClass<SingleModuleClassResolver>()
            )
            field(javaClass<VirtualFileFinder>(),
                  init = GivenExpression(javaClass<VirtualFileFinder>().getName() + ".SERVICE.getInstance(project)"))
        }

private fun generatorForLazyResolveWithJava() =
        generator("compiler/frontend.java/src", "org.jetbrains.jet.di", "InjectorForLazyResolveWithJava") {
            parameter(javaClass<Project>())
            parameter(javaClass<GlobalContext>(), useAsContext = true)
            parameter(javaClass<ModuleDescriptorImpl>(), name = "module", useAsContext = true)
            parameter(javaClass<GlobalSearchScope>(), name = "moduleContentScope")
            parameters(
                    javaClass<BindingTrace>(),
                    javaClass<DeclarationProviderFactory>(),
                    javaClass<ModuleClassResolver>()
            )

            publicFields(
                    javaClass<ResolveSession>(),
                    javaClass<JavaDescriptorResolver>()
            )

            field(javaClass<VirtualFileFinder>(),
                  init = GivenExpression(javaClass<VirtualFileFinderFactory>().getName()
                                         + ".SERVICE.getInstance(project).create(moduleContentScope)")
            )
            fields(
                    javaClass<JavaClassFinderImpl>(),
                    javaClass<TraceBasedExternalSignatureResolver>(),
                    javaClass<LazyResolveBasedCache>(),
                    javaClass<TraceBasedErrorReporter>(),
                    javaClass<PsiBasedMethodSignatureChecker>(),
                    javaClass<PsiBasedExternalAnnotationResolver>(),
                    javaClass<JavaPropertyInitializerEvaluatorImpl>(),
                    javaClass<SamConversionResolverImpl>(),
                    javaClass<JavaSourceElementFactoryImpl>(),
                    javaClass<JavaFlexibleTypeCapabilitiesProvider>(),
                    javaClass<LazyResolveToken>()
            )
            field(javaClass<AdditionalCheckerProvider>(),
                  init = GivenExpression(javaClass<JavaDeclarationCheckerProvider>().getName() + ".INSTANCE$"))
        }

private fun generatorForMacro() =
        generator("compiler/frontend/src", "org.jetbrains.jet.di", "InjectorForMacros") {
            parameter(javaClass<Project>())
            parameter(javaClass<ModuleDescriptor>(), useAsContext = true)

            publicField(javaClass<ExpressionTypingServices>())
            publicField(javaClass<ExpressionTypingComponents>())
            publicField(javaClass<CallResolver>())

            field(javaClass<GlobalContext>(), useAsContext = true,
                  init = GivenExpression("org.jetbrains.jet.context.ContextPackage.GlobalContext()"))

            field(javaClass<AdditionalCheckerProvider>(),
                  init = GivenExpression(javaClass<AdditionalCheckerProvider.Empty>().getCanonicalName() + ".INSTANCE$"))
        }

private fun generatorForTests() =
        generator("compiler/tests", "org.jetbrains.jet.di", "InjectorForTests") {
            parameter(javaClass<Project>())
            parameter(javaClass<ModuleDescriptor>(), useAsContext = true)

            publicFields(
                    javaClass<DescriptorResolver>(),
                    javaClass<ExpressionTypingServices>(),
                    javaClass<ExpressionTypingUtils>(),
                    javaClass<TypeResolver>()
            )

            field(javaClass<GlobalContext>(), init = GivenExpression("org.jetbrains.jet.context.ContextPackage.GlobalContext()"),
                  useAsContext = true)

            field(javaClass<AdditionalCheckerProvider>(),
                  init = GivenExpression(javaClass<JavaDeclarationCheckerProvider>().getName() + ".INSTANCE$"))
        }

private fun generatorForBodyResolve() =
        generator("compiler/frontend/src", "org.jetbrains.jet.di", "InjectorForBodyResolve") {
            parameter(javaClass<Project>())
            parameter(javaClass<GlobalContext>(), useAsContext = true)
            parameter(javaClass<BindingTrace>())
            parameter(javaClass<ModuleDescriptor>(), useAsContext = true)
            parameter(javaClass<AdditionalCheckerProvider>())
            parameter(javaClass<PartialBodyResolveProvider>())

            publicField(javaClass<BodyResolver>())
        }

private fun generatorForLazyResolve() =
        generator("compiler/frontend/src", "org.jetbrains.jet.di", "InjectorForLazyResolve") {
            parameter(javaClass<Project>())
            parameter(javaClass<GlobalContext>(), useAsContext = true)
            parameter(javaClass<ModuleDescriptorImpl>(), useAsContext = true)
            parameter(javaClass<DeclarationProviderFactory>())
            parameter(javaClass<BindingTrace>())
            parameter(javaClass<AdditionalCheckerProvider>())

            publicField(javaClass<ResolveSession>())

            field(javaClass<LazyResolveToken>())
        }

private fun generator(
        targetSourceRoot: String,
        injectorPackageName: String,
        injectorClassName: String,
        body: DependencyInjectorGenerator.() -> Unit
) = generator(targetSourceRoot, injectorPackageName, injectorClassName, "org.jetbrains.jet.generators.injectors.InjectorsPackage", body)
